Docs: Add Precision Support Page (#595)
ROCmSoftwarePlatform/hipSPARSE: docs/reference/precision.rst

.. meta::
  :description: hipSPARSE library precision support overview
  :keywords: hipSPARSE, rocSPARSE, ROCm, API, Sparse Linear Algebra, documentation, precision support, data types

.. _hipsparse-precision-support:

********************************************************************
hipSPARSE precision support
********************************************************************

This section provides an overview of the numerical precision types supported by the hipSPARSE library for sparse
matrix operations.

Supported precision types
=========================

hipSPARSE supports the following precision types across its functions:

.. list-table::
    :header-rows: 1

    *

>       - Type Prefix
>       - C++ Type

Reviewer: ```suggestion
      - Type prefix
      - C++ type
```

the parameter types. For example:

.. code-block:: c

    hipsparseStatus_t hipsparseSgemvi(hipsparseHandle_t handle,
                                      hipsparseOperation_t transA,
                                      int m,
                                      int n,
                                      const float *alpha,
                                      const float *A,
                                      int lda,
                                      int nnz,
                                      const float *x,
                                      const int *xInd,
                                      const float *beta,
                                      float *y,
                                      hipsparseIndexBase_t idxBase,
                                      void *pBuffer)

The parameter types (``float``, ``double``, ``hipFloatComplex``, or ``hipDoubleComplex``)
correspond to the function precision type character and indicate the precision used by that specific function
variant.

Real versus complex precision
-----------------------------

Most sparse matrix operations in hipSPARSE support both real and complex precisions, however certain algorithms

> or optimizations may be specific to real or complex data:

Reviewer: ```suggestion
Most sparse matrix operations in hipSPARSE support both real and complex precisions, however, certain algorithms
or optimizations might be specific to real or complex data:
```

> * Some functions may have different performance characteristics for complex data compared to real data.
> * Certain conversion routines may operate differently depending on whether the data is real or complex.

Reviewer: ```suggestion
* Some functions might have different performance characteristics for complex data compared to real data.
* Certain conversion routines might operate differently depending on whether the data is real or complex.
```

Core precision types
--------------------

The four core precision types (s, d, c, z) are supported across most functions in hipSPARSE. Some specialized

> functions may only support a subset of these types. Always check the specific function documentation to confirm

Reviewer: ```suggestion
functions might only support a subset of these types. See the specific function documentation to confirm
```

variant.

Generic functions and mixed precision
-------------------------------------

hipSPARSE provides generic functions that allow for mixed precision operations. These functions use data type
enumerations to specify precision when creating matrix/vector descriptors and during computation.

For example, when creating a sparse matrix descriptor:

.. code-block:: c

    // Create CSR matrix with single-precision values
    hipsparseCreateCsr(&matA, m, n, nnz,
                       rowPtr, colInd, values,
                       HIPSPARSE_INDEX_32I, HIPSPARSE_INDEX_32I,
                       HIPSPARSE_INDEX_BASE_ZERO, HIP_R_32F);

Or when creating a dense vector:

.. code-block:: c

    // Create dense vector with double-precision values
    hipsparseCreateDnVec(&vecX, size, xValues, HIP_R_64F);

Functions like ``hipsparseSpMV`` use a separate parameter for computation precision:

.. code-block:: c

    // Matrix-vector multiplication with computation in single precision
    hipsparseSpMV(handle, transA, &alpha, matA, vecX, &beta, vecY,
                  HIP_R_32F, HIPSPARSE_MV_ALG_DEFAULT, buffer);

This approach enables:

> * Using different precision types for matrices and vectors.

Reviewer 2: Could change this to:

"_Using different precision types for matrices and vectors. The advantage of using different data types is to save on memory bandwidth and storage when a user application allows while performing the actual computation in a higher precision_"

This helps explain why a user might want to do this. I am fine with not including this as well. Ill defer to your expertise on this. Otherwise looks good. Approved here and in the other PR's.

Reviewer 2: Since this is also related to the point below, you could also just add this addtional comment at the end and a non-bullet point sentence.

Reviewer 2: Something like:

* Using different precision types for matrices and vectors.
* Specifying computation precision independently of storage precision.
* Supporting mixed precision workflows with a unified API.

The advantage of using different data types is to save on memory bandwidth and storage when a user application allows while performing the actual computation in a higher precision

Author: I like the last one, so I integrated that into this PR and the other PR
